src/ert/scheduler/slurm_driver.py

```diff
@@ -0,0 +1,401 @@
+from __future__ import annotations
+
+import asyncio
+import itertools
+import logging
+import shlex
+import stat
+import tempfile
+import time
+from dataclasses import dataclass
+from enum import Enum, auto
+from pathlib import Path
+from typing import (
+    Iterator,
+    Optional,
+    Tuple,
+)
+
+from ert.scheduler.driver import Driver
+from ert.scheduler.event import Event, FinishedEvent, StartedEvent
+
+logger = logging.getLogger(__name__)
+
+
+class JobStatus(Enum):
+    PENDING = auto()
+    COMPLETED = auto()
+    RUNNING = auto()
+    FAILED = auto()
+    CANCELED = auto()
+    COMPLETING = auto()
+    CONFIGURING = auto()
+
+
+@dataclass
+class JobData:
+    iens: int
+    submitted_timestamp: float
+    exit_code: Optional[int] = None
+    status: Optional[JobStatus] = None
+
+
+END_STATES = {JobStatus.FAILED, JobStatus.COMPLETED, JobStatus.CANCELED}
+
+
+@dataclass
+class JobInfo:
+    status: Optional[JobStatus] = None
+
+
+@dataclass
+class ScontrolInfo(JobInfo):
+    exit_code: Optional[int] = None
+
+
+@dataclass
+class SqueueInfo(JobInfo):
+    pass
+
+
+class SlurmDriver(Driver):
+    def __init__(
+        self,
+        exclude_hosts: str = "",
+        include_hosts: str = "",
+        squeue_cmd: str = "squeue",
+        scontrol_cmd: str = "scontrol",
+        scancel_cmd: str = "scancel",
+        sbatch_cmd: str = "sbatch",
+        user: Optional[str] = None,
+        memory: Optional[str] = None,
+        partition: Optional[str] = None,
+        memory_per_cpu: Optional[str] = None,
+        max_runtime: Optional[str] = None,
+        squeue_timeout: Optional[str] = None,
+    ) -> None:
+        super().__init__()
+        self._submit_locks = {}
+        self._iens2jobid = {}
+        self._jobs = {}
+        self._job_error_message_by_iens = {}
+        self._memory_per_cpu = memory_per_cpu
+        self._memory = memory
+        self._max_runtime = max_runtime
+        self._partition = partition
+
+        self._exclude_hosts = exclude_hosts
+        self._include_hosts = include_hosts
+
+        self._sbatch = sbatch_cmd
+        self._sbatch_retries = 1
+
+        self._scancel = scancel_cmd
+        self._squeue = squeue_cmd
+
+        self._scontrol = scontrol_cmd
+        self._scontrol_cache_timestamp = 0
+        self._scontrol_required_cache_age = 30
+        self._scontrol_cache = {}
+
+        self._user = user
+
+        self._sleep_time_between_cmd_retries = 3
+        self._sleep_time_between_kills = 30
+        self._poll_period = 10 if squeue_timeout is None else int(squeue_timeout)
+
+    def _submit_cmd(
+        self,
+        name: str = "dummy",
+        runpath: Optional[Path] = None,
+        num_cpu: Optional[int] = 1,
+    ):
+        sbatch_with_args: list[str] = [
+            str(self._sbatch),
+            f"--job-name={name}",
+            f"-D={runpath}",
```

Comment:
we can use --chdir instead of -D to make it more readable.

Reply:
done